File: languages/go/internal/test/end_to_end_test.go
package test

import (
	"testing"

	"github.com/bloock/bloock-sdk-go/v2/builder"
	"github.com/bloock/bloock-sdk-go/v2/client"
	"github.com/bloock/bloock-sdk-go/v2/client/entity"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestEndToEnd(t *testing.T) {
	sdk := GetSdk()

	t.Run("E2E using all the builders", func(t *testing.T) {
		records := []entity.Record{}

		records = append(records, testFromString(t))
		records = append(records, testFromBytes(t))
		records = append(records, testFromHex(t))
		records = append(records, testFromJson(t))
		records = append(records, testFromFile(t))
		records = append(records, testEcdsaSignature(t, sdk))

		testFromHostedLoader(t)
		testFromIpfsLoader(t)

		testAesEncryption(t)
		testAesEncryptionHosted(t)
		testAesEncryptionIpfs(t)

		testRsaEncryption(t, sdk)
		testRsaEncryptionHosted(t, sdk)
		testRsaEncryptionIpfs(t, sdk)

		testEciesEncryption(t, sdk)
		testEciesEncryptionHosted(t, sdk)
		testEciesEncryptionIpfs(t, sdk)

		testSetProof(t, sdk)

		receipt, err := sdk.SendRecords(records)
		require.NoError(t, err)
		assert.Greater(t, len(receipt), 0)
		assert.NotEqual(t, entity.RecordReceipt{}, receipt[0])

		anchor, err := sdk.WaitAnchor(receipt[0].Anchor, entity.NewAnchorParams())
		require.NoError(t, err)
		assert.Equal(t, receipt[0].Anchor, anchor.Id)

		proof, err := sdk.GetProof(records)
		require.NoError(t, err)
		assert.NotEqual(t, entity.Proof{}, proof)

		root, err := sdk.VerifyProof(proof)
		require.NoError(t, err)
		assert.NotEqual(t, "", proof)

		timestampValidateRoot, err := sdk.ValidateRoot(root, entity.ListOfNetworks().BloockChain)
		require.NoError(t, err)
		assert.Greater(t, timestampValidateRoot, uint64(0))

		network := entity.NewNetworkParams()
		network.Network = entity.ListOfNetworks().BloockChain

		timestampVerifyRecords, err := sdk.VerifyRecords(records, network)
		require.NoError(t, err)
		assert.Greater(t, timestampVerifyRecords, uint64(0))

		assert.Equal(t, timestampValidateRoot, timestampVerifyRecords)
	})
}

func testFromString(t *testing.T) entity.Record {
	record, err := builder.NewRecordBuilderFromString("Hello world").Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd", hash)

	return record
}

func testFromHostedLoader(t *testing.T) {
	payload := "Hello world"
	record, err := builder.NewRecordBuilderFromString(payload).Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	result, err := record.Publish(entity.NewHostedPublisher())
	require.NoError(t, err)
	assert.Equal(t, hash, result)

	record, err = builder.NewRecordBuilderFromLoader(entity.NewHostedLoader(result)).Build()
	require.NoError(t, err)

	hash, err = record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, result, hash)

	assert.Equal(t, payload, string(record.Retrieve()))
}

func testFromIpfsLoader(t *testing.T) {
	payload := "Hello world"
	record, err := builder.NewRecordBuilderFromString(payload).Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	result, err := record.Publish(entity.NewIpfsPublisher())
	require.NoError(t, err)
	assert.Equal(t, hash, result)

	record, err = builder.NewRecordBuilderFromLoader(entity.NewIpfsLoader(result)).Build()
	require.NoError(t, err)

	hash, err = record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, result, hash)

	assert.Equal(t, payload, string(record.Retrieve()))
}

func testFromBytes(t *testing.T) entity.Record {
	record, err := builder.NewRecordBuilderFromBytes([]byte{1, 2, 3, 4, 5}).Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, "7d87c5ea75f7378bb701e404c50639161af3eff66293e9f375b5f17eb50476f4", hash)

	return record
}

func testFromHex(t *testing.T) entity.Record {
	record, err := builder.NewRecordBuilderFromHex("1234567890abcdef").Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, "ed8ab4fde4c4e2749641d9d89de3d920f9845e086abd71e6921319f41f0e784f", hash)

	return record
}

func testFromJson(t *testing.T) entity.Record {
	record, err := builder.NewRecordBuilderFromJSON("{\"hello\":\"world\"}").Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, "586e9b1e1681ba3ebad5ff5e6f673d3e3aa129fcdb76f92083dbc386cdde4312", hash)

	return record
}

func testFromFile(t *testing.T) entity.Record {
	record, err := builder.NewRecordBuilderFromFile([]byte{2, 3, 4, 5, 6}).Build()
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)

	assert.Equal(t, "507aa5dd7b2e52180b764db13c8289ed204109cafe2ef4e453366da8654dc446", hash)

	return record
}

func testAesEncryption(t *testing.T) {
	payload := "Hello world 2"
	password := "some_password"
	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewAesEncrypter(password)).
		Build()

	require.NoError(t, err)
	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	alg, err := encryptedRecord.GetEncryptionAlg()
	assert.Equal(t, entity.AES256GCM, alg)

	_, err = builder.NewRecordBuilderFromRecord(encryptedRecord).
		WithDecrypter(entity.NewAesDecrypter("incorrect_password")).
		Build()
	require.Error(t, err)

	decryptedRecord, err := builder.NewRecordBuilderFromRecord(encryptedRecord).
		WithDecrypter(entity.NewAesDecrypter(password)).
		Build()

	require.NoError(t, err)
	assert.Equal(t, payload, string(decryptedRecord.Retrieve()))

	hash, err := decryptedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testAesEncryptionHosted(t *testing.T) {
	payload := "Hello world 2"
	password := "some_password"

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewAesEncrypter(password)).
		Build()

	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	result, err := encryptedRecord.Publish(entity.NewHostedPublisher())
	require.NoError(t, err)

	loadedRecord, err := builder.NewRecordBuilderFromLoader(entity.NewHostedLoader(result)).
		WithDecrypter(entity.NewAesDecrypter(password)).
		Build()
	require.NoError(t, err)

	assert.Equal(t, payload, string(loadedRecord.Retrieve()))

	hash, err := loadedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testAesEncryptionIpfs(t *testing.T) {
	payload := "Hello world 2"
	password := "some_password"

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewAesEncrypter(password)).
		Build()

	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	result, err := encryptedRecord.Publish(entity.NewIpfsPublisher())
	require.NoError(t, err)

	loadedRecord, err := builder.NewRecordBuilderFromLoader(entity.NewIpfsLoader(result)).
		WithDecrypter(entity.NewAesDecrypter(password)).
		Build()
	require.NoError(t, err)

	require.NoError(t, err)
	assert.Equal(t, payload, string(loadedRecord.Retrieve()))

	hash, err := loadedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testRsaEncryption(t *testing.T, sdk client.Client) {
	payload := "Hello world 2"
	keypair, err := sdk.GenerateRsaKeyPair()

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewRsaEncrypter(keypair.PublicKey)).
		Build()

	require.NoError(t, err)
	assert.NotEqual(t, payload, string(encryptedRecord.Payload))

	alg, err := encryptedRecord.GetEncryptionAlg()
	assert.Equal(t, entity.RSA, alg)

	record, err := builder.NewRecordBuilderFromRecord(encryptedRecord).
		WithDecrypter(entity.NewRsaDecrypter(keypair.PrivateKey)).
		Build()

	require.NoError(t, err)
	assert.Equal(t, payload, string(record.Retrieve()))

	hash, err := record.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testRsaEncryptionHosted(t *testing.T, sdk client.Client) {
	payload := "Hello world 2"
	keypair, err := sdk.GenerateRsaKeyPair()

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewRsaEncrypter(keypair.PublicKey)).
		Build()

	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	result, err := encryptedRecord.Publish(entity.NewHostedPublisher())
	require.NoError(t, err)

	loadedRecord, err := builder.NewRecordBuilderFromLoader(entity.NewHostedLoader(result)).
		WithDecrypter(entity.NewRsaDecrypter(keypair.PrivateKey)).
		Build()
	require.NoError(t, err)

	assert.Equal(t, payload, string(loadedRecord.Retrieve()))

	hash, err := loadedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testRsaEncryptionIpfs(t *testing.T, sdk client.Client) {
	payload := "Hello world 2"
	keypair, err := sdk.GenerateRsaKeyPair()

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewRsaEncrypter(keypair.PublicKey)).
		Build()

	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	result, err := encryptedRecord.Publish(entity.NewIpfsPublisher())
	require.NoError(t, err)

	loadedRecord, err := builder.NewRecordBuilderFromLoader(entity.NewIpfsLoader(result)).
		WithDecrypter(entity.NewRsaDecrypter(keypair.PrivateKey)).
		Build()
	require.NoError(t, err)

	assert.Equal(t, payload, string(loadedRecord.Retrieve()))

	hash, err := loadedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testEciesEncryption(t *testing.T, sdk client.Client) {
	payload := "Hello world 2"
	keypair, err := sdk.GenerateEciesKeyPair()

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewEciesEncrypter(keypair.PublicKey)).
		Build()

	require.NoError(t, err)
	assert.NotEqual(t, payload, string(encryptedRecord.Payload))

	alg, err := encryptedRecord.GetEncryptionAlg()
	assert.Equal(t, entity.ECIES, alg)

	record, err := builder.NewRecordBuilderFromRecord(encryptedRecord).
		WithDecrypter(entity.NewEciesDecrypter(keypair.PrivateKey)).
		Build()

	require.NoError(t, err)
	assert.Equal(t, payload, string(record.Retrieve()))

	hash, err := record.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testEciesEncryptionHosted(t *testing.T, sdk client.Client) {
	payload := "Hello world 2"
	keypair, err := sdk.GenerateEciesKeyPair()

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewEciesEncrypter(keypair.PublicKey)).
		Build()

	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	result, err := encryptedRecord.Publish(entity.NewHostedPublisher())
	require.NoError(t, err)

	loadedRecord, err := builder.NewRecordBuilderFromLoader(entity.NewHostedLoader(result)).
		WithDecrypter(entity.NewEciesDecrypter(keypair.PrivateKey)).
		Build()
	require.NoError(t, err)

	assert.Equal(t, payload, string(loadedRecord.Retrieve()))

	hash, err := loadedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testEciesEncryptionIpfs(t *testing.T, sdk client.Client) {
	payload := "Hello world 2"
	keypair, err := sdk.GenerateEciesKeyPair()

	encryptedRecord, err := builder.NewRecordBuilderFromString(payload).
		WithEncrypter(entity.NewEciesEncrypter(keypair.PublicKey)).
		Build()

	assert.NotEqual(t, payload, string(encryptedRecord.Retrieve()))

	result, err := encryptedRecord.Publish(entity.NewIpfsPublisher())
	require.NoError(t, err)

	loadedRecord, err := builder.NewRecordBuilderFromLoader(entity.NewIpfsLoader(result)).
		WithDecrypter(entity.NewEciesDecrypter(keypair.PrivateKey)).
		Build()
	require.NoError(t, err)

	assert.Equal(t, payload, string(loadedRecord.Retrieve()))

	hash, err := loadedRecord.GetHash()
	require.NoError(t, err)
	assert.Equal(t, "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6", hash)
}

func testEcdsaSignature(t *testing.T, sdk client.Client) entity.Record {
	keys, err := sdk.GenerateKeys()
	require.NoError(t, err)

	name := "Some name"

	record, err := builder.
		NewRecordBuilderFromString("Hello world 3").
		WithSigner(entity.NewEcdsaSigner(entity.SignerArgs{
			PrivateKey: keys.PrivateKey,
			CommonName: &name,
		})).
		Build()
	require.NoError(t, err)

	keys, err = sdk.GenerateKeys()
	require.NoError(t, err)

	recordWithMultipleSignatures, err := builder.
		NewRecordBuilderFromRecord(record).
		WithSigner(entity.NewEcdsaSigner(entity.SignerArgs{PrivateKey: keys.PrivateKey})).
		Build()

	require.NoError(t, err)

	signatures, err := recordWithMultipleSignatures.GetSignatures()
	require.NoError(t, err)
	assert.Equal(t, len(signatures), 2)

	retrievedName, err := signatures[0].GetCommonName()
	assert.Equal(t, name, retrievedName)

	assert.Equal(t, entity.ECDSA, signatures[0].GetAlg())

	return recordWithMultipleSignatures
}

func testSetProof(t *testing.T, sdk client.Client) {
	record, err := builder.NewRecordBuilderFromString("Hello world").Build()
	require.NoError(t, err)

	expectedHash := "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd"

	originalProof := entity.Proof{
		Leaves: []string{expectedHash},
		Nodes:  []string{"ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd"},
		Depth:  "1010101",
		Bitmap: "0101010",
		Anchor: entity.ProofAnchor{
			AnchorID: 42,
			Networks: []entity.AnchorNetwork{{
				Name:   "Ethereum",
				State:  "state",
				TxHash: "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd",
			}},
			Root:   "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd",
			Status: "success",
		},
	}

	err = record.SetProof(originalProof)
	require.NoError(t, err)

	hash, err := record.GetHash()
	require.NoError(t, err)
	assert.Equal(t, expectedHash, hash)

	finalProof, err := sdk.GetProof([]entity.Record{record})
	require.NoError(t, err)

	assert.Equal(t, originalProof, finalProof)
}

File: languages/js/test/end_to_end.test.ts
import {
  Record,
  RecordBuilder,
  BloockClient,
  HostedPublisher,
  HostedLoader,
  AesEncrypter,
  AesDecrypter,
  EcdsaSigner,
  Network,
  RsaEncrypter,
  RsaDecrypter,
  EciesEncrypter,
  EciesDecrypter,
  Proof,
  ProofAnchor,
  AnchorNetwork,
  IpfsPublisher,
  IpfsLoader,
  EnsSigner,
  EncryptionAlg,
  SignatureAlg
} from "../dist/index";
import { describe, test, expect } from "@jest/globals";
import { getSdk } from "./util";

describe("E2E Tests", () => {
  test("test_basic_e2e", async () => {
    let sdk = getSdk();

    const records: Record[] = [];
    records.push(await testFromString());
    records.push(await testFromBytes());
    records.push(await testFromHex());
    records.push(await testFromJson());
    records.push(await testFromFile());
    records.push(await testEcdsaSignature(sdk));
    records.push(await testEnsSignature(sdk));

    await testFromHostedLoader();
    await testFromIpfsLoader();

    await testAesEncryption();
    await testAesEncryptionHosted();
    await testAesEncryptionIpfs();

    await testRsaEncryption(sdk);
    await testRsaEncryptionHosted(sdk);
    await testRsaEncryptionIpfs(sdk);

    await testEciesEncryption(sdk);
    await testEciesEncryptionHosted(sdk);
    await testEciesEncryptionIpfs(sdk);

    await testSetProof(sdk);

    const sendReceipt = await sdk.sendRecords(records);
    expect(sendReceipt.length).toBeGreaterThan(0);

    const anchor = await sdk.waitAnchor(sendReceipt[0].anchor);
    expect(anchor.id).toEqual(sendReceipt[0].anchor);

    const proof = await sdk.getProof(records);
    const root = await sdk.verifyProof(proof);

    const timestampValidateRoot = await sdk.validateRoot(
      root,
      Network.BLOOCK_CHAIN
    );
    expect(timestampValidateRoot).toBeGreaterThan(0);

    const timestampVerifyRecords = await sdk.verifyRecords(
      records,
      Network.BLOOCK_CHAIN
    );
    expect(timestampVerifyRecords).toBeGreaterThan(0);

    expect(timestampValidateRoot).toEqual(timestampVerifyRecords);
  }, 120000);
});

async function testFromString(): Promise<Record> {
  let record = await RecordBuilder.fromString("Hello world").build();

  let hash = await record.getHash();
  expect(hash).toEqual(
    "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd"
  );

  return record;
}

async function testFromBytes(): Promise<Record> {
  let record = await RecordBuilder.fromBytes(
    Uint8Array.from([1, 2, 3, 4, 5])
  ).build();

  let hash = await record.getHash();
  expect(hash).toEqual(
    "7d87c5ea75f7378bb701e404c50639161af3eff66293e9f375b5f17eb50476f4"
  );

  return record;
}

async function testFromHex(): Promise<Record> {
  let record = await RecordBuilder.fromHex("1234567890abcdef").build();

  let hash = await record.getHash();
  expect(hash).toEqual(
    "ed8ab4fde4c4e2749641d9d89de3d920f9845e086abd71e6921319f41f0e784f"
  );

  return record;
}

async function testFromJson(): Promise<Record> {
  let record = await RecordBuilder.fromJson({ hello: "world" }).build();

  let hash = await record.getHash();
  expect(hash).toEqual(
    "586e9b1e1681ba3ebad5ff5e6f673d3e3aa129fcdb76f92083dbc386cdde4312"
  );

  return record;
}

async function testFromFile(): Promise<Record> {
  let record = await RecordBuilder.fromFile(
    Uint8Array.from([2, 3, 4, 5, 6])
  ).build();

  let hash = await record.getHash();
  expect(hash).toEqual(
    "507aa5dd7b2e52180b764db13c8289ed204109cafe2ef4e453366da8654dc446"
  );

  return record;
}

async function testEcdsaSignature(sdk: BloockClient): Promise<Record> {
  let keys = await sdk.generateKeys();
  let name = "Some name";

  let record = await RecordBuilder.fromString("Hello world 3")
    .withSigner(new EcdsaSigner(keys.privateKey, { commonName: name }))
    .build();

  keys = await sdk.generateKeys();

  const recordWithMultipleSignatures = await RecordBuilder.fromRecord(record)
    .withSigner(new EcdsaSigner(keys.privateKey))
    .build();

  let signatures = await recordWithMultipleSignatures.getSignatures();
  expect(signatures.length).toEqual(2);

  expect(await signatures[0].getCommonName()).toEqual(name);
  expect(signatures[0].getAlg()).toEqual(SignatureAlg.ECDSA);

  return recordWithMultipleSignatures;
}

async function testEnsSignature(sdk: BloockClient): Promise<Record> {
  let keys = await sdk.generateKeys();

  let record = await RecordBuilder.fromString("Hello world 4")
    .withSigner(new EnsSigner(keys.privateKey))
    .build();

  let signatures = await record.getSignatures();
  expect(signatures.length).toEqual(1);

  signatures[0].signature =
    "66e0c03ce895173be8afac992c43f49d0bea3768c8146b83df9acbaee7e67d7106fd2a668cb9c90edd984667caf9fbcd54acc460fb22ba5e2824eb9811101fc601";
  signatures[0].messageHash =
    "7e43ddd9df3a0ca242fcf6d1b190811ef4d50e39e228c27fd746f4d1424b4cc6";
  expect(await signatures[0].getCommonName()).toEqual("vitalik.eth");
  expect(signatures[0].getAlg()).toEqual(SignatureAlg.ENS);

  return record;
}

async function testFromHostedLoader() {
  let record = await RecordBuilder.fromString("Hello world").build();

  let hash = await record.getHash();

  let result = await record.publish(new HostedPublisher());
  expect(result).toEqual(hash);

  record = await RecordBuilder.fromLoader(new HostedLoader(result)).build();

  hash = await record.getHash();

  expect(hash).toEqual(result);
}

async function testFromIpfsLoader() {
  let payload = "Hello world";
  let record = await RecordBuilder.fromString(payload).build();

  let hash = await record.getHash();

  let result = await record.publish(new IpfsPublisher());
  expect(result).toEqual(hash);

  record = await RecordBuilder.fromLoader(new IpfsLoader(result)).build();

  hash = await record.getHash();

  expect(hash).toEqual(result);
}

async function testAesEncryption() {
  let payload = "Hello world 2";
  let password = "some_password";
  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new AesEncrypter(password))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  expect(await encrypted_record.getEncryptionAlg()).toEqual(
    EncryptionAlg.AES256GCM
  );

  await expect(
    RecordBuilder.fromRecord(encrypted_record).withDecrypter(
      new AesDecrypter("incorrect_password")
    ).build
  ).rejects.toThrow();

  let decrypted_record = await RecordBuilder.fromRecord(encrypted_record)
    .withDecrypter(new AesDecrypter(password))
    .build();

  expect(String.fromCharCode(...decrypted_record.payload)).toEqual(payload);

  let hash = await decrypted_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );

  return hash;
}

async function testAesEncryptionHosted() {
  let payload = "Hello world 2";
  let password = "some_password";
  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new AesEncrypter(password))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  let result = await encrypted_record.publish(new HostedPublisher());

  let loaded_record = await RecordBuilder.fromLoader(new HostedLoader(result))
    .withDecrypter(new AesDecrypter(password))
    .build();

  expect(String.fromCharCode(...loaded_record.payload)).toEqual(payload);

  let hash = await loaded_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );

  return hash;
}

async function testAesEncryptionIpfs() {
  let payload = "Hello world 2";
  let password = "some_password";
  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new AesEncrypter(password))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  let result = await encrypted_record.publish(new IpfsPublisher());

  let loaded_record = await RecordBuilder.fromLoader(new IpfsLoader(result))
    .withDecrypter(new AesDecrypter(password))
    .build();

  expect(String.fromCharCode(...loaded_record.payload)).toEqual(payload);

  let hash = await loaded_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );

  return hash;
}

async function testRsaEncryption(sdk: BloockClient) {
  let payload = "Hello world 2";
  let keypair = await sdk.generateRsaKeyPair();

  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new RsaEncrypter(keypair.publicKey))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  expect(await encrypted_record.getEncryptionAlg()).toEqual(EncryptionAlg.RSA);

  let decrypted_record = await RecordBuilder.fromRecord(encrypted_record)
    .withDecrypter(new RsaDecrypter(keypair.privateKey))
    .build();

  expect(String.fromCharCode(...decrypted_record.payload)).toEqual(payload);

  let hash = await decrypted_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );
}

async function testRsaEncryptionHosted(sdk: BloockClient) {
  let payload = "Hello world 2";
  let keypair = await sdk.generateRsaKeyPair();

  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new RsaEncrypter(keypair.publicKey))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  let result = await encrypted_record.publish(new HostedPublisher());

  let loaded_record = await RecordBuilder.fromLoader(new HostedLoader(result))
    .withDecrypter(new RsaDecrypter(keypair.privateKey))
    .build();

  expect(String.fromCharCode(...loaded_record.payload)).toEqual(payload);

  let hash = await loaded_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );
}

async function testRsaEncryptionIpfs(sdk: BloockClient) {
  let payload = "Hello world 2";
  let keypair = await sdk.generateRsaKeyPair();

  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new RsaEncrypter(keypair.publicKey))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  let result = await encrypted_record.publish(new IpfsPublisher());

  let loaded_record = await RecordBuilder.fromLoader(new IpfsLoader(result))
    .withDecrypter(new RsaDecrypter(keypair.privateKey))
    .build();

  expect(String.fromCharCode(...loaded_record.payload)).toEqual(payload);

  let hash = await loaded_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );
}

async function testEciesEncryption(sdk: BloockClient) {
  let payload = "Hello world 2";
  let keypair = await sdk.generateEciesKeyPair();

  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new EciesEncrypter(keypair.publicKey))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  expect(await encrypted_record.getEncryptionAlg()).toEqual(
    EncryptionAlg.ECIES
  );

  let decrypted_record = await RecordBuilder.fromRecord(encrypted_record)
    .withDecrypter(new EciesDecrypter(keypair.privateKey))
    .build();

  expect(String.fromCharCode(...decrypted_record.payload)).toEqual(payload);

  let hash = await decrypted_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );
}

async function testEciesEncryptionHosted(sdk: BloockClient) {
  let payload = "Hello world 2";
  let keypair = await sdk.generateEciesKeyPair();

  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new EciesEncrypter(keypair.publicKey))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  let result = await encrypted_record.publish(new HostedPublisher());

  let loaded_record = await RecordBuilder.fromLoader(new HostedLoader(result))
    .withDecrypter(new EciesDecrypter(keypair.privateKey))
    .build();

  expect(String.fromCharCode(...loaded_record.payload)).toEqual(payload);

  let hash = await loaded_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );
}

async function testEciesEncryptionIpfs(sdk: BloockClient) {
  let payload = "Hello world 2";
  let keypair = await sdk.generateEciesKeyPair();

  let encrypted_record = await RecordBuilder.fromString(payload)
    .withEncrypter(new EciesEncrypter(keypair.publicKey))
    .build();

  expect(String.fromCharCode(...encrypted_record.payload)).not.toEqual(payload);

  let result = await encrypted_record.publish(new IpfsPublisher());

  let loaded_record = await RecordBuilder.fromLoader(new IpfsLoader(result))
    .withDecrypter(new EciesDecrypter(keypair.privateKey))
    .build();

  expect(String.fromCharCode(...loaded_record.payload)).toEqual(payload);

  let hash = await loaded_record.getHash();
  expect(hash).toEqual(
    "96d59e2ea7cec4915c415431e6adb115e3c0c728928773bcc8e7d143b88bfda6"
  );
}

async function testSetProof(sdk: BloockClient) {
  let record = await RecordBuilder.fromString("Hello world").build();

  let original_proof = new Proof(
    ["ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd"],
    ["ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd"],
    "1010101",
    "0101010",
    new ProofAnchor(
      42,
      [
        new AnchorNetwork(
          "Ethereum",
          "state",
          "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd"
        )
      ],
      "ed6c11b0b5b808960df26f5bfc471d04c1995b0ffd2055925ad1be28d6baadfd",
      "success"
    )
  );

  await record.setProof(original_proof);

  let finalProof = await sdk.getProof([record]);

  expect(finalProof).toEqual(original_proof);
}
